File: solutions/Leetcode_1339/leetcode_1339.cpp
#include <algorithm>
#include <vector>

class Solution1
{
public:
    int maxProduct(TreeNode *root)
    {
        _dfs(root);
        long ret = 0;
        for (const int x : _subSums)
        {
            ret = std::max(ret, x * (_total - x));
        }
        return ret % (int)(1e9 + 7);
    }

private:
    long _total = 0;
    std::vector<int> _subSums;

    int _dfs(TreeNode *node)
    {
        if (node == nullptr)
            return 0;
        _total += node->val;
        int subSum = node->val + _dfs(node->left) + _dfs(node->right);
        _subSums.push_back(subSum);
        return subSum;
    }
};

File: solutions/Leetcode_1339/leetcode_1339.py
from typing import Optional

from leetcode.python.tree import TreeNode


class Solution1:
    def max_product(self, root: Optional[TreeNode]) -> int:
        self._total = 0
        self._sub_sum = []
        self._dfs(root)
        ret = 0
        for x in self._sub_sum:
            ret = max(ret, x * (self._total - x))
        return ret % (10**9 + 7)

    def _dfs(self, node):
        if not node:
            return 0
        self._total += node.val
        sub_sum = node.val + self._dfs(node.left) + self._dfs(node.right)
        self._sub_sum.append(sub_sum)
        return sub_sum
